### PR TITLE
WDPD-166: Guaranteed Invoice by Wirecard Frontend Payment Method

### PR DESCRIPTION
**Pull request by ps-amaureder**
### This PR

* Adds Input fields for date of birth and phone
* Adds select field for saving input data
* Adds mechanism to show/hide payment methods(e.g. hide guaranteed invoice if user < 18)

### Notes

* Depends on wdpd-178
* if basket is empty it is not possible to get to the payment method select step in oxid
* Guaranteed Invoice by Wirecard payment method is only testable with country set to Austria, to be able to use Austria add it in backend (Master Settings->Countries)

### TODO

* Add check for digital goods
* Add check for currency
* Add check for shipping country
* Add check for billing country
* Handle different shipping and billing address handling

as soon as the configured values are available

* Write Tests

### How to Test

* Check if input fields are correctly shown/hidden to/from user
* Check if save selection is correctly shown/hidden to/from user
* Check if saving/ not saving works correctly
* Check if showing/hiding of payment method works correctly
* Test with logged in and guest user

### Jira Links
https://jira.parkside.at/browse/WDPD-166

### CODE REVIEW

File: Core/PaymentMethodHelper.php [hunk 1]
```diff
@@ -129,6 +136,174 @@ public static function getBic()
         return $aDynvalues['bic'];
     }
 
+    /**
+     * Returns date of birth
+     *
+     * @return string date of birth formated for db (format 'Y-m-d')
```
Comment (ps-holler) [inline Core/PaymentMethodHelper.php]:
```suggestion
     * @return string date of birth formatted for db (format 'Y-m-d')
```

File: translations/de/module_de_lang.php [hunk 2]
```diff
@@ -15,13 +15,22 @@
     'wd_account_holder_title' => 'Kontoinhaber',
     'wd_amount' => 'Betrag',
     'wd_bic' => 'BIC',
+    'wd_birthdate_format_php_code' => 'd.m.Y',
```
Comment (ps-holler) [inline translations/de/module_de_lang.php]: I'd make this more generic and just call it `wd_date_format`.

File: translations/de/module_de_lang.php [hunk 3]
```diff
@@ -15,13 +15,22 @@
...
+    'wd_birthdate_format_user_hint' => 'DD.MM.JJJJ',
```
Comment (ps-holler) [inline translations/de/module_de_lang.php]: Maybe `wd_date_format_placeholder`?

File: Core/PaymentMethodHelper.php [hunk 4]
```diff
@@ -129,6 +136,174 @@ public static function getBic()
...
+    }
+
+    /**
+     * Returns true if user is min 18 years old or date of birth is not known
+     *
+     * @return bool
+     *
+     * @since 1.2.0
+     */
+    public static function isUserEighteen()
```
Comment (ps-holler) [inline Core/PaymentMethodHelper.php]: Can we make this more generic, e.g.:

```suggestion
    public static function isUserOlderThan($iAge)
```

File: Core/PaymentMethodHelper.php [hunk 5]
```diff
@@ -129,6 +136,174 @@ public static function getBic()
...
+    {
+        $oSession = Registry::getConfig()->getSession();
+        $aDynvalues = $oSession->getVariable('dynvalue');
+
+        $oDateOfBirth =
+            DateTime::createFromFormat(Helper::translate('wd_birthdate_format_php_code'), $aDynvalues['dateOfBirth']);
+
+        if (!$oDateOfBirth) {
```
Comment (ps-holler) [inline Core/PaymentMethodHelper.php]: Shouldn't this return `false` if no birthdate is set?

File: Core/PaymentMethodHelper.php [hunk 6]
```diff
@@ -129,6 +136,174 @@ public static function getBic()
...
+    }
+
+    /**
+     * Returns true if a phone number is needed
+     *
+     * @return bool
+     *
+     * @since 1.2.0
+     */
+    public static function isPhoneNeeded()
```
Comment (ps-holler) [inline Core/PaymentMethodHelper.php]: I think this is too specific for a helper.

File: Core/PaymentMethodHelper.php [hunk 7]
```diff
@@ -129,6 +136,174 @@ public static function getBic()
...
+    }
+
+    /**
+     * Sets the saveCheckoutFields flag
+     *
+     * @param int $iSave value 1 if checkout data should be saved 0 if not
+     *
+     * @since 1.2.0
+     */
+    public static function setSaveCheckoutFields($iSave)
```
Comment (ps-holler) [inline Core/PaymentMethodHelper.php]: This would be a perfect candidate for a `SessionHelper`.

File: translations/de/module_de_lang.php (hunk 2 above)
Comment (ps-amaureder) [reply translations/de/module_de_lang.php]: i changed it to wd_date_format_php_code, the php_code part should make it clear for the translator, that this format is needed for the plugin to work correctly

File: translations/de/module_de_lang.php (hunk 3 above)
Comment (ps-amaureder) [reply translations/de/module_de_lang.php]: I changed it to wd_date_format_user_hint

File: Core/PaymentMethodHelper.php (hunk 7 above)
Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: I will create a SessionHelper

File: Core/PaymentMethodHelper.php [hunk 8]
```diff
@@ -203,4 +378,53 @@ public static function prepareCreditorName()
 
         return $sCreditorName ? $sCreditorName : $oShop->oxshops__oxcompany;
     }
+
+    /**
+     * Checks the user data if mandatory fields are set correctly for guaranteed invoice and saves them if needed
+     *
+     * @since 1.2.0
+     */
+    public static function checkPayStepUserInput()
```
Comment (ps-cgrach) [inline Core/PaymentMethodHelper.php]: I think we have something like that for eps and giropay. Is there somehting in the controller for this!?

File: Core/PaymentMethodHelper.php [hunk 9]
```diff
@@ -203,4 +378,53 @@ public static function prepareCreditorName()
...
+
+        self::_validateUserInput();
+    }
+
+    /**
+     * Validates the user input and redirects to the payment step with an error if needed
+     *
+     * @since 1.2.0
+     */
+    private static function _validateUserInput()
```
Comment (ps-cgrach) [inline Core/PaymentMethodHelper.php]: see comment above

File: Core/PaymentMethodHelper.php (hunk 6 above)
Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: yes I will change it

Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: it is not needed for gi by wirecard, I removed it and will handle it in gi by payolution in the payment method itself

File: Core/PaymentMethodHelper.php (hunk 8 above)
Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: not yet, but I guess we have to refactor it later

File: Core/PaymentMethodHelper.php (hunk 9 above)
Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: will need refactoring after everything is merged

File: Core/PaymentMethodHelper.php (hunk 8 above)
Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: moved to payment method

File: Core/PaymentMethodHelper.php (hunk 9 above)
Comment (ps-amaureder) [reply Core/PaymentMethodHelper.php]: moved to payment methdod

**Pull request merged**